woocommerce/woocommerce-ios
# [Mobile Payments] Add local mobile discovery to StripeCardReaderService

Closes: #8172


## Description


In sTAP Away, we will add the ability to take card payments using the phone's built in card reader. This PR adds `LocalMobileReaderDelegate` conformance to our `StripeCardReaderService`, and discovery method support to our calls into the `CardReaderService`.

The Tap on Mobile functionality is behind a feature flag, and only available on `localDeveloper` Xcode builds. Alpha and Release builds do not currently have the correct entitlement to use the built-in reader functionality, and the feature toggle is not available on those builds.

When the feature toggle is enabled, all connection attempts will be made to the built-in reader; when it's disabled, all attempts will be bluetooth. In future, the feature flag will be changed to show the user-facing selection screen.

This PR does not address any UI changes, it only adds the feature flag, and the ability to connect to the built-in reader for payments when the toggle is enabled.

### Duplication

There's quite a bit of duplicated code in the `StripeCardReaderService`. Future work will address this, for now it's important to keep the changes as simple as possible to enable us to work on the user-facing flows sooner.

## Testing instructions


### Simulated reader

The Stripe simulated reader can be used if this does not build to a device with your installed provisioning profiles. Updates to the profiles have been requested.

To enable the Stripe simulated reader, in Xcode go to `Product > Scheme > Edit Scheme` and select `Run` in the sidebar, then `Arguments` in the tabs. Check the `-simulate-stripe-card-reader` box.

### Testing a 

1. Using a WCPay or Stripe US store
2. Navigate to `Menu > Settings cog > Experimental features`
3. Toggle `Tap to Pay on iPhone`
4. Go back to `Menu`
5. Tap `Payments > Collect payment` and go through the Simple Payments flow.
6. Select `Card` on the payment methods screen, and connect to the (built-in) reader
7. Observe that you can take a payment using the built in reader.

## Screenshots


### With simulated reader

Note that there is a bug in the Stripe simulated reader, which means the `Preparing reader` screen remains on screen and the other screens do not show the payment details.

https://user-images.githubusercontent.com/2472348/203127504-aae44b89-b95a-4f25-8eb6-97d040d5ad82.mp4



### With physical device

This shows the flow, but the screen recording does not capture the out-of-process Apple Tap on iPhone payment screen, which is shown over the `Reader is ready` screen you see in this video.


https://user-images.githubusercontent.com/2472348/203127549-c01e6867-62aa-4496-8d38-99c438d75567.mp4


### Simulator without turning on the simulated reader

If you try to use the built-in (non-simulated) reader in a simulator, you'll see this error

![tap-on-iphone-without-simulated-reader](https://user-images.githubusercontent.com/2472348/203127619-0bc481ce-486d-4143-801e-2e216eb40eef.jpg)


---
